python/ray/experimental/packaging/load_package.py

```diff
@@ -0,0 +1,226 @@
+"""Support for loading code packages into Ray at runtime.
+
+Ray packages allow developers to define self-contained code modules that can
+be imported reproducibly into any Ray cluster. Each package can define its own
+runtime environment, which can include:
+ - Different versions of code (e.g., from different git commits).
+ - Different Python libraries (e.g., conda environments, pip dependencies).
+ - Different Docker container images.
+
+You can run this file for an example of loading a "hello world" package.
+"""
+
+import importlib.util
+import os
+import re
+import hashlib
+import subprocess
+import tempfile
+import yaml
+
+import ray
+import ray._private.runtime_env as runtime_support
+
+
+def load_package(config_path: str) -> "_RuntimePackage":
+    """Load the code package given its config path.
+
+    Args:
+        config_path (str): The path to the configuration YAML that defines
+            the package. For documentation on the packaging format, see the
+            example YAML in ``example_pkg/ray_pkg.yaml``.
+
+    Examples:
+        >>> # Load from local.
+        >>> my_pkg = load_package("~/path/to/my_pkg.yaml")
+
+        >>> # Load from GitHub.
+        >>> my_pkg = ray.util.load_package(
+        ...   "https://raw.githubusercontent.com/user/repo/refspec"
+        ...   "/path/to/package/my_pkg.yaml")
+
+        >>> # Inspect the package runtime env.
+        >>> print(my_pkg._runtime_env)
+        ... {"conda": {...},
+        ...  "docker": "anyscale-ml/ray-ml:nightly-py38-cpu",
+        ...  "files": "https://github.com/demo/foo/blob/v3.0/project/"}
+
+        >>> # Run remote functions from the package.
+        >>> my_pkg.my_func.remote(1, 2)
+
+        >>> # Create actors from the package.
+        >>> actor = my_pkg.MyActor.remote(3, 4)
+
+        >>> # Create new remote funcs in a package.
+        >>> @ray.remote(runtime_env=my_pkg._runtime_env)
+        >>> def f(): ...
+    """
+
+    if not ray.is_initialized():
+        # TODO(ekl) lift this requirement?
+        raise RuntimeError("Ray must be initialized first to load packages.")
+
+    config_path = _download_from_github_if_needed(config_path)
+
+    if not os.path.exists(config_path):
+        raise ValueError("Config file does not exist: {}".format(config_path))
+
+    # TODO(ekl) validate schema?
+    config = yaml.safe_load(open(config_path).read())
+    base_dir = os.path.abspath(os.path.dirname(config_path))
+    runtime_env = config["runtime_env"]
+
+    # Autofill working directory by uploading to GCS storage.
```

Comment:
This part is a little unclear to me. For code that's local, we'd still need to package it. For GH repos, it may be advantageous to download it once and ship the small package, rather than have every node try to download from GH independently (which might lead to rate limiting, ec.)